components/camel-aws2-athena/src/main/java/org/apache/camel/component/aws2/athena/Athena2QueryHelper.java

```diff
@@ -0,0 +1,452 @@
+    this.waitTimeout = determineWaitTimeout(exchange, configuration);
+    this.delay = determineDelay(exchange, configuration);
+    this.maxAttempts = determineMaxAttempts(exchange, configuration);
+    this.retry = determineRetry(exchange, configuration);
+    this.resetWaitTimeoutOnAttempt = determineResetWaitTimeoutOnRetry(exchange, configuration);
+    this.absoluteStartMs = now();
+
+    this.currentDelay = determineInitialDelay(exchange, configuration);
+  }
+
+  private long now() {
+    return clock.millis();
+  }
+
+  long getElapsedMillis() {
+    return now() - this.absoluteStartMs;
+  }
+
+  /**
+   * Record that a query attempt was made.  This is relevant b/c only so many attempts are permitted.
+   */
+  void markAttempt() {
+    if (attempts == 0) {
+      this.startMs = now();
+    } else {
+      if (resetWaitTimeoutOnAttempt) {
+        this.startMs = now();
+      }
+    }
+    ++attempts;
+
+    this.isFailure = false;
+    this.isSuccess = false;
+    this.isRetry = false;
+  }
+
+  int getAttempts() {
+    return this.attempts;
+  }
+
+  /**
+   * Should another query attempt be made?
+   */
+  boolean shouldAttempt() {
+    if (this.attempts >= this.maxAttempts) {
+      LOG.trace("AWS Athena start query execution used all {} attempts", this.maxAttempts);
+      return false;
+    }
+
+    if (this.interrupted) {
+      LOG.trace("AWS Athena start query execution thread was interrupted, will try no more");
+      return false;
+    }
+
+    if (this.isFailure) {
+      LOG.trace("AWS Athena start query execution detected permanent failure");
+      return false;
+    }
+
+    if (this.isSuccess) {
+      LOG.trace("AWS Athena start query execution detected success, will try no more");
+      return false;
+    }
+
+    // if this.isRetry, return true
+
+    return true;
+  }
+
+  /**
+   * Should there be a wait for the query to complete?
+   */
+  boolean shouldWait() {
+    long now = now();
+    long millisWaited = now - this.startMs;
+    if (millisWaited >= this.waitTimeout) {
+      LOG
+          .trace("AWS Athena start query execution waited for {}, which exceeded wait timeout of {}", millisWaited,
+              this.waitTimeout);
+      return false;
+    }
+
+    if (this.interrupted) {
+      LOG.trace("AWS Athena start query execution thread was interrupted, will wait no longer");
+      return false;
+    }
+
+    if (this.isFailure) {
+      LOG.trace("AWS Athena start query execution detected failure, will wait no longer");
+      return false;
+    }
+
+    if (this.isSuccess) {
+      LOG.trace("AWS Athena start query execution detected success, will wait no longer");
+      return false;
+    }
+
+    if (this.isRetry) {
+      LOG.trace("AWS Athena start query execution detected retry, will immediately attempt retry");
+      return false;
+    }
+
+    return true;
+  }
+
+  void doWait() {
+    try {
+      Thread.sleep(this.currentDelay);
```

Comment:
Is this the 'right' way of doing this in a Camel component?

Reply:
Personally I think it's ok, because with Athena we need to wait for results, but lets wait for more feedback @davsclaus @gnodet